### PR TITLE
(WIP; DO NOT REVIEW) Use python tensor subclass version of NT for PT2

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99753

I'm out next week so I'm dumping some stuff here for whoever is interested. Added some notes inline. Other things:
* Main idea: convert real NT -> fake / meta instance of `NestedTensor` python tensor subclass during tracing and use that throughout the rest of the PT2 stack
* Python tensor subclass only supports contiguous for now
    * Only need `sizes` because of this, and `sizes` is a dim-length list containing int / List[int] items (for ragged dims) OR SymInts for dynamic shapes (will probably only support this case anyway)
* Inference-only for now; ignore Autograd
* Skip functionalization on NTs for now (there's no in-place op support anyway, although we should ideally handle NT <-> T views within functionalization)

Example script:
```python
import torch
from torch.nested import nested_tensor
from torch.nested._nested_tensor import NestedTensor
from torch._inductor import debug

torch._inductor.config.debug = True
torch._dynamo.config.traceable_tensor_subclasses.add(NestedTensor)

device = 'cuda'

def make_tensor(*shape, device=device, dtype=torch.float32):
    return torch.randn(*shape, device=device, dtype=dtype)

torch.manual_seed(1)

def fn(x, x_offsets):
    x_nt = torch._nested_view_from_jagged(x, x_offsets)
    x_nt = x_nt + 69
    x_nt = x_nt * 42
    return x_nt

torch._dynamo.disallow_in_graph(torch.diff)

compiled_fn = torch.compile(fn)

# shape (sum(*), D)
# component shapes: (3, 5), (4, 5), (6, 5)
x = make_tensor(13, 5)
x_offsets = torch.tensor([0, 3, 7, 13], dtype=torch.int64, device=device)

# helps create dynamic graph right away
torch._dynamo.mark_dynamic(x, 0)
torch._dynamo.mark_dynamic(x_offsets, 0)

output = compiled_fn(x, x_offsets)

# shape (sum(*), D)
# component shapes: (2, 5), (6, 5), (4, 5), (5, 5)
y = make_tensor(17, 5)
y_offsets = torch.tensor([0, 2, 8, 12, 17], dtype=torch.int64, device=device)

output2 = compiled_fn(y, y_offsets)

print(output)
print(output2)
```

WARNING: AWFUL HACKS AHEAD. DO NOT REVIEW YET.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @soumith @anijain2305 @desertfire

### CODE REVIEW

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -1119,6 +1119,88 @@ def __eq__(self, other: object) -> bool:
         return False
 
 
+def _automatic_dynamic(e, tx, name, static_shapes):
```
Comment (jbschlosser) [inline torch/_dynamo/variables/builder.py:1122]: all this shit was factored out from `wrap_to_fake_tensor_and_record()`; no changes

File: torch/_functorch/aot_autograd.py [hunk 2]
```diff
@@ -630,7 +630,7 @@ def gen_alias_from_base(aliased_base_tensor, target_meta_tensor, target_requires
     return aliased_out
 
 def to_fun(t):
-    if isinstance(t, Tensor):
+    if isinstance(t, Tensor) and not t.is_nested:
```
Comment (jbschlosser) [inline torch/_functorch/aot_autograd.py:633]: Avoid trying to functionalize nested tensors within AOTAutograd for now

File: torch/_subclasses/fake_tensor.py [hunk 3]
```diff
@@ -810,8 +810,18 @@ def should_allow_numbers_as_tensors(func: OpOverload):
 class FakeTensorConfig:
     debug = os.environ.get("TORCH_FAKE_TENSOR_DEBUG", False)
 
+# Hack to make isinstance(x, FakeTensor) return True when x is a fake-ified NestedTensor.
+# A NestedTensor that has been fake-ified simply has a fake buffer.
```
Comment (jbschlosser) [inline torch/_subclasses/fake_tensor.py:814]: self explanatory hopefully

File: torch/_subclasses/meta_utils.py [hunk 4]
```diff
@@ -265,6 +266,49 @@ def sym_sizes_strides_storage_offset(t):
                         with torch.enable_grad():
                             r = r.clone()
                             r._coalesced_(t.is_coalesced())
+                elif t.is_nested:
```
Comment (jbschlosser) [inline torch/_subclasses/meta_utils.py:269]: start of the logic for converting real NT -> meta / fake instance of NestedTensor python subclass

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -1684,30 +1686,36 @@ def create_symbol(
...
+                # Small performance optimization: if we have a min-max constraint,
+                # we can proactively narrow to that range
+                if isinstance(constraint_dim, StrictMinMaxConstraint):
+                    assert not duck
+                    self.var_to_range[sympy_expr] &= constraint_dim.vr
+
+                vr = self.var_to_range[sympy_expr]
+                if val not in vr:
+                    raise RuntimeError(f"{val} not in range [{vr.lower}, {vr.upper}]")
+            elif isinstance(val, tuple):
```
Comment (jbschlosser) [inline torch/fx/experimental/symbolic_shapes.py:1705]: Hacking in support for tuple-backed SymInts using `sympy.Array`?? (tons more changes needed to do this right)

NB: tuple here because lists aren't hashable but tuples are

File: torch/nested/_nested_tensor.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        return padded
+
+    def __getattribute__(self, key):
+        if key == "is_nested":
+            return True
+        if key == "shape":
+            return self.size()
+
+        # These are hacks to get around the fact that the NestedTensor itself is not a FakeTensor,
+        # but simply contains a FakeTensor buffer.
```
Comment (jbschlosser) [inline torch/nested/_nested_tensor.py:133]: also self explanatory

File: torch/_tensor_str.py [hunk 7]
```diff
@@ -527,11 +527,18 @@ def _str_intern(inp, *, tensor_contents=None):
...
+                if self.dtype != torch.get_default_dtype():
+                    suffixes.append("dtype=" + str(self.dtype))
+                if not custom_contents_provided:
+                    tensor_str = "..."
+            else:
+                strs = ",\n".join(
+                    indented_str(str(t), indent + 1)
+                    for t in torch.ops.aten.unbind.int(self, 0)
+                )
+                tensor_str = f"[\n{strs}\n]"
```
Comment (jbschlosser) [inline torch/_tensor_str.py:541]: leftover from old approach; this isn't called if we use python `NestedTensor` exclusively for meta / fake NTs. Rather, `NestedTensor.__repr__()` will be called.

File: torch/nested/_nested_tensor.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+        s = self.unbind()
+        list_str = '\n'.join([str(t) for t in s])
+        return f"{prefix}([\n{list_str}\n])"
+
+    def dim(self):
+        return self.total_dim
+
+    # nt.shape calls here as well
+    def size(self, dim=None):
```
Comment (jbschlosser) [inline torch/nested/_nested_tensor.py:85]: NB: probably need `stride()` too for downstream, and we can easily auto-generate it from `sizes` because of contiguous constraint

File: torch/_dynamo/variables/builder.py (hunk 1 above)
Comment (ezyang) [reply torch/_dynamo/variables/builder.py:1122]: We're modifying this code a lot, recommend carving this into its own PR and landing ASAP.

File: torch/nested/_nested_tensor.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+# Properties:
+#  * Constrained to contiguous-only
+#  * Returned for fake / meta conversion
+#  * Supports size() / stride() -> [SymInt] for dynamic shapes
+#  * Supports size() / stride() -> [Union[int, List[int]]] for non-dynamic shapes
+class NestedTensor(Tensor):
+    @staticmethod
+    def __new__(cls, buffer, *args, **kwargs):
+        # Put in obvious garbage; we don't ever want to actually hit the C++ side of this thing
+        return Tensor._make_subclass(cls, torch.empty(0))
```
Comment (ezyang) [inline torch/nested/_nested_tensor.py:22]: you probably want to make the device on the model tensor correct

File: torch/nested/_nested_tensor.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+
+    # sizes should be a list of SymInts for the dynamic shapes case
+    def __init__(self, buffer: Tensor, sizes: List[Union[int, List[int], SymInt]]):
+        self.buffer = buffer
```
Comment (ezyang) [inline torch/nested/_nested_tensor.py:26]: doc for buffer format would be handy for readers

Comment (ezyang) [reply torch/nested/_nested_tensor.py:26]: You don't want the ragged sizes to be a List[int] right? Isn't it going to be a tensor?

File: torch/_dynamo/variables/builder.py (hunk 1 above)
Comment (jbschlosser) [reply torch/_dynamo/variables/builder.py:1122]: #101114

File: torch/_functorch/aot_autograd.py [hunk 11]
```diff
@@ -672,7 +672,7 @@ def to_fun(t):
         if isinstance(t, Tensor):
             if t in memo:
                 return memo[t]
-            r = torch._to_functional_tensor(t, mirror_autograd_meta=True)
+            r = torch._to_functional_tensor(t, mirror_autograd_meta=True) if not t.is_nested else t
```
Comment (bdhirsh) [inline torch/_functorch/aot_autograd.py:675]: let me know if/when you want some help with this - I think in theory it shouln't be too hard to get functionalization working (there are probably some view ops specific to nested tensor, that we need to define "inverses" for)

Comment (jbschlosser) [reply torch/_functorch/aot_autograd.py:675]: thanks! for now, I'm avoiding it to get benchmark results but I will definitely need your help at some point :)
